app/src/main/java/com/zulip/android/activities/RecyclerMessageAdapter.java

```diff
@@ -45,5 +46,50 @@
         mDefaultStreamHeaderColor = ContextCompat.getColor(context, R.color.stream_header);
         mDefaultPrivateMessageColor = ContextCompat.getColor(context, R.color.huddle_body);
         privateHuddleText = context.getResources().getString(R.string.huddle_text);
+        setupLists(messageList);
+    }
+    private int getHeaderIndex(String id) {
+        for (int i = 0; i < items.size(); i++) {
+            if (items.get(i) instanceof MessageHeaderParent) {
+                MessageHeaderParent item = (MessageHeaderParent) items.get(i);
+                if (item.getId().equals(id)) {
+                    return i;
+                }
+            }
+        }
+        return -1;
+    }
+
+    private void setupLists(List<Message> messageList) {
+        for (Message message : messageList) {
+            addMessage(message);
+        }
+    }
+
+    void addMessage(Message message) {
+
+        int index = getHeaderIndex(message.getIdForHolder());
+
+        if (index < 0) { //No messageParent for this one
+            MessageHeaderParent messageHeaderParent = new MessageHeaderParent((message.getStream() == null) ? null : message.getStream().getName(), message.getSubject(), message.getIdForHolder());
+            messageHeaderParent.setMessageType(message.getType());
+            messageHeaderParent.setDisplayRecipent(message.getDisplayRecipient(zulipApp));
+            if (message.getType() == MessageType.STREAM_MESSAGE)
```

Comment:
This looks like it's missing the `{ }` for the if statement. I know syntactically this is ok, but this block is a bit hard to read. It'd be more concise with them in this case.

Generally, any if statement that spans multiple lines should have brackets.